<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Описание-данных" data-toc-modified-id="Описание-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Описание данных</a></span></li><li><span><a href="#Цели-исследования" data-toc-modified-id="Цели-исследования-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Цели исследования</a></span></li><li><span><a href="#Загрузим-данные" data-toc-modified-id="Загрузим-данные-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Загрузим данные</a></span></li><li><span><a href="#Задачи" data-toc-modified-id="Задачи-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Задачи</a></span><ul class="toc-item"><li><span><a href="#Посчитайте,-сколько-книг-вышло-после-1-января-2000-года" data-toc-modified-id="Посчитайте,-сколько-книг-вышло-после-1-января-2000-года-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Посчитайте, сколько книг вышло после 1 января 2000 года</a></span></li><li><span><a href="#Для-каждой-книги-посчитайте-количество-обзоров-и-среднюю-оценку" data-toc-modified-id="Для-каждой-книги-посчитайте-количество-обзоров-и-среднюю-оценку-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Для каждой книги посчитайте количество обзоров и среднюю оценку</a></span></li><li><span><a href="#Определите-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-—-так-вы-исключите-из-анализа-брошюры;" data-toc-modified-id="Определите-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-—-так-вы-исключите-из-анализа-брошюры;-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;</a></span></li><li><span><a href="#Определите-автора-с-самой-высокой-средней-оценкой-книг-—-учитывайте-только-книги-с-50-и-более-оценками;" data-toc-modified-id="Определите-автора-с-самой-высокой-средней-оценкой-книг-—-учитывайте-только-книги-с-50-и-более-оценками;-4.4"><span class="toc-item-num">4.4&nbsp;&nbsp;</span>Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;</a></span></li><li><span><a href="#Посчитайте-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-48-оценок." data-toc-modified-id="Посчитайте-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-48-оценок.-4.5"><span class="toc-item-num">4.5&nbsp;&nbsp;</span>Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок.</a></span></li></ul></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Вывод</a></span></li></ul></div>

## Описание данных

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

## Цели исследования

Наша компания купила крупный сервис для чтения книг по подписке. Цель исследования — проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

## Загрузим данные

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import text, create_engine

# устанавливаем параметры
db_config = {'user': '###', # имя пользователя
'pwd': '###', # пароль
'host': '###',
'port': ###, # порт подключения
'db': '#####'} # название базы данных
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)

# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

In [2]:
# чтобы выполнить SQL-запрос, используем Pandas
for table in ['books', 'authors', 'publishers', 'ratings', 'reviews']:
    query = '''SELECT * FROM {} LIMIT 5'''.format(table)
    con=engine.connect()
    display(pd.io.sql.read_sql(sql=text(query), con = con))

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

## Задачи

### Посчитайте, сколько книг вышло после 1 января 2000 года

In [3]:
query = '''
SELECT COUNT(book_id) AS books_published
FROM books
WHERE publication_date > '2000-01-01'
'''
con=engine.connect()

pd.io.sql.read_sql(sql=text(query), con = con)

books_published
0              819

После 1 января 2000 года вышло 819 книг, то есть за 20 лет - это, примерно, по 41 книге в год

### Для каждой книги посчитайте количество обзоров и среднюю оценку

In [4]:
query = '''
SELECT 
--b.book_id AS book_id,
title,
COUNT(DISTINCT r.review_id) AS reviews_cnt,
AVG(ra.rating) AS avg_rating
FROM books as b
LEFT JOIN reviews AS r on b.book_id = r.book_id
LEFT JOIN ratings AS ra on b.book_id = ra.book_id
GROUP BY b.book_id
ORDER BY reviews_cnt, avg_rating DESC
'''
con=engine.connect()

pd.io.sql.read_sql(sql=text(query), con = con)

title  reviews_cnt  \
0     The Cat in the Hat and Other Dr. Seuss Favorites            0   
1                                 Leonardo's Notebooks            0   
2                            Essential Tales and Poems            0   
3    Disney's Beauty and the Beast (A Little Golden...            0   
4      Anne Rice's The Vampire Lestat: A Graphic Novel            0   
..                                                 ...          ...   
995                                      The Alchemist            6   
996                                           The Road            6   
997                          The Giver (The Giver  #1)            6   
998                                    Eat  Pray  Love            6   
999                            Twilight (Twilight  #1)            7   

     avg_rating  
0      5.000000  
1      4.000000  
2      4.000000  
3      4.000000  
4      3.666667  
..          ...  
995    3.789474  
996    3.772727  
997    3.750000  
998    3.395833  
999    3.662500  

[1000 rows x 3 columns]

Самые высокие оценки у книг, на которые нет обзоров, самые низкие у книг с наибольшим количеством обзоров. Возможно, это обусловлено тем, что недовольные читатели чаще хотят объяснить низкую оценку и дают отзыв, довольные просто ставят оценку. Сумерки лидируют по количеству обзоров, но имеют низкий рейтинг.

### Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;

In [5]:
query = '''
SELECT 
pb.publisher,
tab.published_books
FROM
(
SELECT
publisher_id,
COUNT(book_id) AS published_books
FROM books
WHERE book_id in
(SELECT
book_id
FROM books
WHERE num_pages > 50)
GROUP BY publisher_id
ORDER BY published_books DESC
LIMIT 1
) AS tab
LEFT JOIN publishers AS pb ON tab.publisher_id = pb.publisher_id
'''
con=engine.connect()

pd.io.sql.read_sql(sql=text(query), con = con)

publisher  published_books
0  Penguin Books               42

Издательство Penguin Books выпустило наибольшее число книг толще 50 страниц

### Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;

In [6]:
query = '''
SELECT
author,
ROUND(avg_rating) AS avg_rating
FROM
(
    SELECT
    b.author_id,
    AVG(tab.avg_rating) AS avg_rating
    FROM
(
        SELECT
        book_id,
        AVG(rating) AS avg_rating,
        COUNT(rating_id) AS rating_cnt
        FROM ratings
        GROUP BY book_id
        HAVING COUNT(rating_id) >= 50
        ORDER BY avg_rating DESC
) AS tab
    LEFT JOIN books AS b ON tab.book_id = b.book_id
    GROUP BY b.author_id
    ORDER BY avg_rating DESC
    LIMIT 1
) AS tab2
LEFT JOIN authors AS a ON tab2.author_id = a.author_id


'''
con=engine.connect()

pd.io.sql.read_sql(sql=text(query), con = con)

author  avg_rating
0  J.K. Rowling/Mary GrandPré         4.0

Книги J.K. Rowling о Гарри Поттере имеют самую высокую среднюю оценку среди книг с 50 и более оценками. Это неудивительно: эти книги популярны по всему миру.

### Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

In [7]:
query = '''
SELECT
AVG(review_cnt) AS avg_review_cnt

FROM
(
    SELECT 
    username,
    COUNT(review_id) AS review_cnt
    FROM reviews
    WHERE username IN
(
        SELECT
        username
        FROM ratings
        GROUP BY username
        HAVING COUNT(rating_id) > 48
)
    GROUP BY username
    ) AS tab
   
'''
con=engine.connect()

pd.io.sql.read_sql(sql=text(query), con = con)

avg_review_cnt
0            24.0

В среднем пользователи, которые поставили больше 48 оценок, оставляют 24 обзора. 

## Вывод

Мы проанализировали базу данных с целью сформулировать предложение для сервиса для чтения книг по подписке. Исследование показало, что после 1 января 2000 года вышло 819 книг, то есть за 20 лет - это, примерно, по 41 книге в год. Самые высокие оценки у книг, на которые нет обзоров, самые низкие у книг с наибольшим количеством обзоров. Возможно, это обусловлено тем, что недовольные читатели чаще хотят объяснить низкую оценку и дают отзыв, довольные просто ставят оценку. Сумерки лидируют по количеству обзоров, но имеют низкий рейтинг.  
Среди издательств лидирует издательство Penguin Books, которое выпустило наибольшее число книг толще 50 страниц. А самую высокую среднюю оценку среди книг с 50 и более оценками имеют книги J.K. Rowling о Гарри Поттере. Это неудивительно: эти книги популярны по всему миру.  Также мы узнали, что в среднем пользователи, которые поставили больше 48 оценок, оставляют по 24 обзора.